src/app/clusters/groups-server/groups-server.cpp

```diff
@@ -60,107 +65,120 @@
-    if (status != EMBER_ZCL_STATUS_SUCCESS)
-    {
-        emberAfGroupsClusterPrintln("ERR: writing name support %x", status);
-    }
+    return groups->GroupMappingExists(fabricIndex, mapping);
 }
 
-// --------------------------
-// Internal functions used to maintain the group table within the context
-// of the binding table.
-//
-// In the binding:
-// The first two bytes of the identifier is set to the groupId
-// The local endpoint is set to the endpoint that is mapped to this group
-// --------------------------
-static EmberAfStatus addEntryToGroupTable(EndpointId endpoint, GroupId groupId, const CharSpan & groupName)
+static EmberAfStatus addEntryToGroupTable(FabricIndex fabricIndex, EndpointId endpointId, GroupId groupId,
+                                          const CharSpan & groupName)
 {
-    uint8_t i;
+    VerifyOrReturnError(IsFabricGroupId(groupId), EMBER_ZCL_STATUS_INVALID_VALUE);
 
     // Check for duplicates.
-    if (isGroupPresent(endpoint, groupId))
+    if (isGroupPresent(fabricIndex, endpointId, groupId))
     {
         // Even if the group already exists, tell the application about the name,
         // so it can cope with renames.
-        emberAfPluginGroupsServerSetGroupNameCallback(endpoint, groupId, groupName);
         return EMBER_ZCL_STATUS_DUPLICATE_EXISTS;
     }
 
-    // Look for an empty binding slot.
-    for (i = 0; i < EMBER_BINDING_TABLE_SIZE; i++)
+    GroupDataProvider * groups = GetGroupDataProvider();
+    VerifyOrReturnError(nullptr != groups, EMBER_ZCL_STATUS_NOT_FOUND);
+    GroupDataProvider::GroupMapping mapping(endpointId, groupId, groupName);
+
+    CHIP_ERROR err = groups->AddGroupMapping(fabricIndex, mapping);
+    return (CHIP_NO_ERROR == err) ? EMBER_ZCL_STATUS_SUCCESS : EMBER_ZCL_STATUS_INSUFFICIENT_SPACE;
+}
+
+static EmberAfStatus removeEntryFromGroupTable(FabricIndex fabricIndex, EndpointId endpointId, GroupId groupId)
+{
+    VerifyOrReturnError(IsFabricGroupId(groupId), EMBER_ZCL_STATUS_INVALID_VALUE);
+
+    if (isGroupPresent(fabricIndex, endpointId, groupId))
     {
-        EmberBindingTableEntry binding;
-        if (emberGetBinding(i, &binding) == EMBER_SUCCESS && binding.type == EMBER_UNUSED_BINDING)
+        GroupDataProvider * groups = GetGroupDataProvider();
+        VerifyOrReturnError(nullptr != groups, EMBER_ZCL_STATUS_NOT_FOUND);
+        GroupDataProvider::GroupMapping mapping(endpointId, groupId);
+
+        CHIP_ERROR err = groups->RemoveGroupMapping(fabricIndex, mapping);
+        if (CHIP_NO_ERROR == err)
+        {
+            return EMBER_ZCL_STATUS_SUCCESS;
+        }
+        else
         {
-            EmberStatus status;
-            binding.type    = EMBER_MULTICAST_BINDING;
-            binding.groupId = groupId;
-            binding.local   = endpoint;
+            emberAfGroupsClusterPrintln("ERR: Failed to remove mapping (end:0x%x, group:0x%x)", endpointId, groupId);
+            return EMBER_ZCL_STATUS_FAILURE;
```

Comment:
If we're going to bury the CHIP_ERROR returned by RemoveGroupMapping in the translation to EmberAfStatus, can we at least log this in emberAfGroupsClusterPrintln?

Reply:
Yes we can. I have added the error code to the log.